File: JAVA_Templates/StringHashing.java
package JAVA_Templates;

import java.math.BigInteger;
import java.util.Random;

class Hashing {
    long[] hash1, hash2;
    long[] inv1, inv2;
    int n;
    static int muresiplier = 43;
    static final Random rnd = new Random();
    static final int mod1 =
        BigInteger.valueOf((int)(1e9 + rnd.nextInt((int) 1e9)))
        .nextProbablePrime()
        .intValue();
    static final int mod2 =
        BigInteger.valueOf((int)(1e9 + rnd.nextInt((int) 1e9)))
        .nextProbablePrime()
        .intValue();
    static final int invMuresiplier1 =
        BigInteger.valueOf(muresiplier).modInverse(BigInteger.valueOf(mod1)).intValue();
    static final int invMuresiplier2 =
        BigInteger.valueOf(muresiplier).modInverse(BigInteger.valueOf(mod2)).intValue();

    public Hashing(String s) {
        n = s.length();
        hash1 = new long[n + 1];
        hash2 = new long[n + 1];
        inv1 = new long[n + 1];
        inv2 = new long[n + 1];
        inv1[0] = 1;
        inv2[0] = 1;
        long p1 = 1;
        long p2 = 1;
        for (int i = 0; i < n; i++) {
            hash1[i + 1] = (hash1[i] + s.charAt(i) * p1) % mod1;
            p1 = p1 * muresiplier % mod1;
            inv1[i + 1] = inv1[i] * invMuresiplier1 % mod1;
            hash2[i + 1] = (hash2[i] + s.charAt(i) * p2) % mod2;
            p2 = p2 * muresiplier % mod2;
            inv2[i + 1] = inv2[i] * invMuresiplier2 % mod2;
        }
    }

    public long getHash(int i, int len) {
        return (((hash1[i + len] - hash1[i] + mod1) * inv1[i] % mod1) << 32)
               + (hash2[i + len] - hash2[i] + mod2) * inv2[i] % mod2;
    }

    public long getHashbounds(int x, int y) {
        return getHash(x, y - x + 1);
    }
}

File: python_templates/string_hashing.py
import random


class Hashing:
    MULTIPLIER = 43
    MOD1 = 1000000007
    MOD2 = 1000000009

    def __init__(self, s):
        self.n = len(s)
        self.hash1 = [0] * (self.n + 1)
        self.hash2 = [0] * (self.n + 1)
        self.inv1 = [0] * (self.n + 1)
        self.inv2 = [0] * (self.n + 1)
        self.inv1[0] = 1
        self.inv2[0] = 1

        inv_multiplier1 = pow(self.MULTIPLIER, self.MOD1 - 2, self.MOD1)
        inv_multiplier2 = pow(self.MULTIPLIER, self.MOD2 - 2, self.MOD2)

        p1 = 1
        p2 = 1

        for i in range(self.n):
            self.hash1[i + 1] = (self.hash1[i] + ord(s[i]) * p1) % self.MOD1
            p1 = p1 * self.MULTIPLIER % self.MOD1
            self.inv1[i + 1] = self.inv1[i] * inv_multiplier1 % self.MOD1

            self.hash2[i + 1] = (self.hash2[i] + ord(s[i]) * p2) % self.MOD2
            p2 = p2 * self.MULTIPLIER % self.MOD2
            self.inv2[i + 1] = self.inv2[i] * inv_multiplier2 % self.MOD2

    def get_hash(self, i, length):
        h1 = (
            (self.hash1[i + length] - self.hash1[i] + self.MOD1) * self.inv1[i]
        ) % self.MOD1
        h2 = (
            (self.hash2[i + length] - self.hash2[i] + self.MOD2) * self.inv2[i]
        ) % self.MOD2
        return (h1 << 32) + h2

    def get_hash_bounds(self, x, y):
        return self.get_hash(x, y - x + 1)
